syntax = "proto3";
package im;

option cc_enable_arenas = true;

import "auth_service.proto";
import "friend_service.proto";
import "message_service.proto";

// ============================================================================
// Protocol Layer - CommandType and Envelope
// ============================================================================

enum CommandType {
    CMD_UNKNOWN       = 0;

    // Auth commands
    CMD_REGISTER_REQ  = 1;
    CMD_REGISTER_RES  = 2;
    CMD_LOGIN_REQ     = 3;
    CMD_LOGIN_RES     = 4;

    // Friend commands
    CMD_ADD_FRIEND_REQ = 5;
    CMD_ADD_FRIEND_RES = 6;
    CMD_FRIEND_REQ_PUSH = 7;
    CMD_HANDLE_FRIEND_REQ = 8;
    CMD_HANDLE_FRIEND_RES = 9;
    CMD_FRIEND_STATUS_PUSH = 10;
    CMD_GET_FRIEND_LIST_REQ = 11;
    CMD_GET_FRIEND_LIST_RES = 12;

    // Message commands
    CMD_P2P_MSG_REQ   = 50;
    CMD_P2P_MSG_PUSH  = 51;
    CMD_MSG_ACK       = 52;
}

message Envelope {
    uint64 seq = 1;
    CommandType cmd = 2;
    int64 timestamp = 3;

    oneof payload {
        RegisterReq register_req = 100;
        RegisterResp register_res = 101;
        LoginReq login_req = 102;
        LoginResp login_res = 103;

        AddFriendReq add_friend_req = 200;
        AddFriendResp add_friend_res = 201;
        FriendReqPush friend_req_push = 202;
        HandleFriendReq handle_friend_req = 203;
        HandleFriendResp handle_friend_res = 204;
        FriendStatusPush friend_status_push = 205;
        GetFriendListReq get_friend_list_req = 206;
        GetFriendListResp get_friend_list_res = 207;

        P2PMessage p2p_msg_req = 300;
        P2PMessage p2p_msg_push = 301;
        MessageAck msg_ack = 302;
    }
}
